File: week1/p100.cpp
#include<bits/stdc++.h>
using namespace std;

int L[1000000];
int N=100000;

int collatz(int k){
    int res;
    if(k>=1 && k<=N && L[k-1]!=-1)return L[k-1];
    if(k%2==1)res=collatz(3*k+1)+1;
    else res=collatz(k/2)+1;
    if(k>=1 && k<=N)L[k-1]=res;
    return res;
}

int main(){
    int a,b,i,ans;
    L[0]=1;
    for(i=1;i<1000000;i++)L[i]=-1;
    while(cin>>a>>b){
        ans=0;
        for(i=min(a,b);i<=max(a,b);i++)ans=max(ans,collatz(i));
        cout<<a<<" "<<b<<" "<<ans<<"\n";
    }
}

File: 2020/week1/p100.py
memo=[-1 for i in range(1000000)]
memo[0]=1
N=100000

def collatz(k):
    if k>=1 and k<=N:
        if memo[k-1]!=-1:
            return memo[k-1]
    
    if k%2==1:
        res=collatz(3*k+1)+1
    else:
        res=collatz(k//2)+1
    
    if k>=1 and k<=N:
        memo[k-1]=res
    return res

while True:
    try:
        In=list(map(int,input().split()))
        a=In[0]
        b=In[1]
        ans=0
        for i in range(min(a,b),max(a,b)+1):
            ans=max(ans,collatz(i))
        print(a,b,ans)
    except EOFError: break